### ISSUE TITLE
LiteDB crash issues

### ISSUE THREAD

**Issue by Quicksilv3r710** [StrangeLoopGames/EcoIssues#5343]
Keep getting this error a lot, file is in use by another process/database is locked. Causes the server to crash.  Seems to happen more often when there is more people logged in.
[Crash LiteException 03024446.txt](https://github.com/StrangeLoopGames/EcoIssues/files/1777677/Crash.LiteException.03024446.txt)

**Comment by Quicksilv3r710** [comment]
Every time the browser is opened, within 30 seconds it crashes with "file is in use by another process when attempting to backup"

**Comment by MarjieAVolk** [comment]
```
LiteDB.LiteException: Timeout. Database is locked for more than 00:01:00.
   at LiteDB.LockService.Write()
   at LiteDB.LiteEngine.Transaction[T](String collection, Boolean addIfNotExists, Func`2 action)
   at LiteDB.LiteEngine.Insert(String collection, BsonDocument doc, BsonType autoId)
   at LiteDB.LiteCollection`1.Insert(T document)
   at Eco.Simulation.WorldLayers.WorldLayerStats.<>c__DisplayClass0_0.<CollectStats>b__0(AnimalLayer layer)
   at Eco.Shared.Utils.EnumerableExtensions.ForEach[T](IEnumerable`1 enumeration, Action`1 action)
   at Eco.Simulation.WorldLayers.WorldLayerStats.CollectStats()
   at Eco.Simulation.EcoSim.CollectStats()
   at Eco.Core.Plugins.TickTimeUtil.TimeSubprocess(Action func)
   at Eco.Simulation.EcoSim.DoTick(TickSample tick)
   at Eco.Simulation.EcoSim.Run()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart(Object obj)
```

**Comment by MarjieAVolk** [comment]
Others report this crash happens while the server is doing its backup

**Comment by pamduncan** [comment]
We had no crashes after disabling stats with this db lockup issue but players wanted to try it again so we turned it back on this morning.  WIthin 20 minutes we were experiencing crashes again.

--BEGIN DUMP--
Dump Time
03/13/2018 13:28:42

Exception
LiteDB.LiteException: Timeout. Database is locked for more than 00:01:00.
   at LiteDB.LockService.Read()
   at LiteDB.LiteEngine.<Find>d__9.MoveNext()
   at LiteDB.LiteEngine.<Find>d__13.MoveNext()
   at LiteDB.LiteCollection`1.<Find>d__17.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Linq.Enumerable.LastOrDefault[TSource](IEnumerable`1 source)
   at Eco.Stats.ContinuousPlayerStatManager`1.RecordAction(T action)
   at Eco.Core.Utils.AtomicAction.SimpleAtomicAction.TryApply()
   at Eco.Core.Utils.AtomicAction.MultiAtomicAction.TryApply()
   at Eco.Gameplay.Players.User.Tick()
   at Eco.Simulation.Simulation.TickAll()
   at Eco.Gameplay.Players.UserManager.Run()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart(Object obj)

--END DUMP--

******************************************

--BEGIN DUMP--
Dump Time
03/13/2018 13:28:42

Exception
LiteDB.LiteException: Timeout. Database is locked for more than 00:01:00.
   at LiteDB.LockService.Read()
   at LiteDB.LiteEngine.<Find>d__9.MoveNext()
   at LiteDB.LiteEngine.<Find>d__13.MoveNext()
   at LiteDB.LiteCollection`1.<Find>d__17.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Linq.Enumerable.LastOrDefault[TSource](IEnumerable`1 source)
   at Eco.Stats.ContinuousPlayerStatManager`1.RecordAction(T action)
   at Eco.Core.Utils.AtomicAction.SimpleAtomicAction.TryApply()
   at Eco.Core.Utils.AtomicAction.MultiAtomicAction.TryApply()
   at Eco.Gameplay.Players.User.Tick()
   at Eco.Simulation.Simulation.TickAll()
   at Eco.Gameplay.Players.UserManager.Run()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart(Object obj)

--END DUMP--


We have had to turn stats back off to stop the crashing

**Comment by MarjieAVolk** [comment]
Save file that saw this: https://www.dropbox.com/s/oi316fy96wei2x3/Game.zip?dl=0

**Comment by SaiyerDed** [comment]
Also having db lockups. Server threw three logs at me within several seconds of eachother. I'll put them in the order I received them in:



--BEGIN DUMP--

Dump Time

03/15/2018 10:34:13

 

Exception

System.Exception: Caught exception invoking RPC PlayerInteract on Player!

 

Timeout. Database is locked for more than 00:01:00.

 

 

Stack:

   at LiteDB.LockService.Write()

   at LiteDB.LiteEngine.Transaction[T](String collection, Boolean addIfNotExists, Func`2 action)

   at LiteDB.LiteEngine.Insert(String collection, BsonDocument doc, BsonType autoId)

   at LiteDB.LiteCollection`1.Insert(T document)

   at Eco.Core.Utils.AtomicAction.SimpleAtomicAction.TryApply()

   at Eco.Core.Utils.AtomicAction.MultiAtomicAction.TryApply()

   at Eco.Core.Utils.AtomicAction.MultiAtomicAction.TryApply()

   at Eco.Gameplay.Items.UsableItemUtils.PlayerDeleteBlock(Vector3i blockPosition, Player player, Boolean addToInventory, BlockItem fallbackGiveItem, IAtomicAction[] additionalActions)

   at Eco.Gameplay.Items.ToolItem.PlayerDeleteBlock(Vector3i blockPosition, Player player, Boolean addToInventory, Single calorieMultiplier, BlockItem fallbackGiveItem, IAtomicAction[] additionalActions)

   at Eco.Mods.TechTree.ShovelItem.OnActLeft(InteractionContext context)

   at Eco.Gameplay.Interactions.InteractionExtensions.ExecuteUntilSuccess(IEnumerable`1 results)

   at Eco.Gameplay.Players.Player.PlayerInteract(InteractionInfo info)

 

   at Eco.Shared.Networking.RPCManager.TryInvoke(Object controller, String methodname, BSONObject bsonArgs, Object& result)

   at Eco.Shared.Networking.RPCManager.InvokeOn(BSONObject bson, Object controller, String methodname)

   at Eco.Shared.Networking.RPCManager.HandleReceiveRPC(INetClient client, BSONObject bson)

   at Eco.Shared.Networking.NetPort.ProcessBuffer()

   at Eco.Shared.Networking.NetPort.Update()

   at Eco.Plugins.Networking.Clients.Client.<.ctor>b__56_0()

 

--END DUMP--



--BEGIN DUMP--

Dump Time

03/15/2018 10:34:14

 

Exception

System.Exception: Caught exception invoking RPC ViewRPC on Client!

 

Caught exception invoking RPC IsRelevant on Law!

 

Timeout. Database is locked for more than 00:01:00.

 

 

Stack:

   at LiteDB.LockService.Read()

   at LiteDB.LiteEngine.<Find>d__9.MoveNext()

   at LiteDB.LiteEngine.<Find>d__13.MoveNext()

   at LiteDB.LiteCollection`1.<Find>d__17.MoveNext()

   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()

   at System.Linq.Enumerable.Sum(IEnumerable`1 source)

   at Eco.Gameplay.LegislationSystem.LawLogic.Conditions.NumericComparisonCondition.IsMet(IPlayerAction action)

   at Eco.Gameplay.LegislationSystem.LawLogic.ControlFlow.LawControlFlowList.IsRelevant(IPlayerAction action)

   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)

   at System.Linq.Enumerable.All[TSource](IEnumerable`1 source, Func`2 predicate)

 

 

 

Stack:

   at Eco.Shared.Networking.RPCManager.TryInvoke(Object controller, String methodname, BSONObject bsonArgs, Object& result)

   at Eco.Shared.Networking.RPCManager.InvokeOn(BSONObject bson, Object controller, String methodname)

   at Eco.Plugins.Networking.Clients.Client.ViewRPC(Int32 id, String methodname, BSONObject bson)

 

   at Eco.Shared.Networking.RPCManager.TryInvoke(Object controller, String methodname, BSONObject bsonArgs, Object& result)

   at Eco.Shared.Networking.RPCManager.InvokeOn(BSONObject bson, Object controller, String methodname)

   at Eco.Shared.Networking.RPCManager.HandleReceiveRPC(INetClient client, BSONObject bson)

   at Eco.Shared.Networking.NetPort.ProcessBuffer()

   at Eco.Shared.Networking.NetPort.Update()

   at Eco.Plugins.Networking.Clients.Client.<.ctor>b__56_0()

 

--END DUMP--



--BEGIN DUMP--

Dump Time

03/15/2018 10:34:17

 

Exception

LiteDB.LiteException: Timeout. Database is locked for more than 00:01:00.

   at LiteDB.LockService.Read()

   at LiteDB.LiteEngine.<Find>d__9.MoveNext()

   at LiteDB.LiteEngine.<Find>d__13.MoveNext()

   at LiteDB.LiteCollection`1.<Find>d__17.MoveNext()

   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()

   at System.Linq.Enumerable.LastOrDefault[TSource](IEnumerable`1 source)

   at Eco.Stats.ContinuousPlayerStatManager`1.RecordAction(T action)

   at Eco.Core.Utils.AtomicAction.SimpleAtomicAction.TryApply()

   at Eco.Core.Utils.AtomicAction.MultiAtomicAction.TryApply()

   at Eco.Gameplay.Players.User.Tick()

   at Eco.Simulation.Simulation.TickAll()

   at Eco.Gameplay.Players.UserManager.Run()

   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)

   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)

   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)

   at System.Threading.ThreadHelper.ThreadStart(Object obj)

 

--END DUMP--

**Comment by Last2Night** [comment]
7.2.5 crashes with the database lockup exception

`--BEGIN DUMP--
Dump Time
03/16/2018 03:38:33

Exception
LiteDB.LiteException: Timeout. Database is locked for more than 00:01:00.
   at LiteDB.LockService.Read()
   at LiteDB.LiteEngine.<Find>d__9.MoveNext()
   at LiteDB.LiteEngine.<Find>d__13.MoveNext()
   at LiteDB.LiteCollection`1.<Find>d__17.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Linq.Enumerable.LastOrDefault[TSource](IEnumerable`1 source)
   at Eco.Stats.ContinuousPlayerStatManager`1.RecordAction(T action)
   at Eco.Core.Utils.AtomicAction.SimpleAtomicAction.TryApply()
   at Eco.Core.Utils.AtomicAction.MultiAtomicAction.TryApply()
   at Eco.Gameplay.Components.AirPollutionComponent.Tick()
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at Eco.Gameplay.Objects.WorldObject.Tick()
   at Eco.Shared.Utils.EnumerableExtensions.ForEach[T](IEnumerable`1 enumeration, Action`1 action)
   at Eco.Gameplay.Objects.WorldObjectManager.Run()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart(Object obj)

--END DUMP--
`

**Comment by SaiyerDed** [comment]
Looks like my server just went crazy over someone starting world leader voting and votes coming in which resulted in a pretty bad db lockup.

What my dump folder showed: https://i.imgur.com/NLh03Di.png
ZIP containing the above dumps: https://www.dropbox.com/s/pe38r0jf69e906r/Eco%20Dumps.rar?dl=0

**Comment by Last2Night** [comment]
In 7.2.5 this has become a major issue. It crashes our server every 30 minutes at least, even though i have set up backup to once per 2 hours. Since 7.2.5 deployment this is the only crash reason ive seen but its extremely frequent and interrupts backup process, our server can easily be missing 6-10 hours of backups due to this issue.

**Comment by metachronism** [comment]
@Last2Night @SaiyerDed could either of you try out latest staging and see if the crashes are far less frequent?  and/or link your game.eco/game.db files here?

**Comment by pamduncan** [comment]
More Crash and exception files re this issue:  https://www.dropbox.com/s/su9havpjtyaaar2/Crash%20and%20Exception%20Files%20-DB%20Lock.zip?dl=0

**Comment by Last2Night** [comment]
@metachronism Switching to staging version might make users very unhappy, however I have no problem providing you with the db files. Where do you want me to upload them to? Take in consideration that they are slightly over 500MB in size.

**Comment by MarjieAVolk** [comment]
If you can zip them up and just post them here, that would be great.  Or if they are too big, you can upload them to a file sharing service such as dropbox or google drive.

**Comment by Last2Night** [comment]
@metachronism 
https://drive.google.com/file/d/1cRPyKK4OJxlJ6Z4y0duAIi7lXIUJ-Svr/view?usp=sharing

**Comment by SaiyerDed** [comment]
I believe my post over at #6174 contains the zips you're looking for. I've had more since then, but they all revolve around the same thing. Let me know if you'd like the most recent ones as well.

As for moving to staging, I'm in the same boat as @Last2Night where my community might not all hop on board for the staging version. Especially since not all of them are in our Discord server to communicate the change.

**Comment by MarjieAVolk** [comment]
Anybody who's still having trouble with this, you can get around it for now by disabling stats on your server.  There is a setting for it in the server window.  This should hopefully be a lot better on 7.3 when it does get officially released.

**Comment by pamduncan** [comment]
More crashes with db_lockups
https://www.dropbox.com/s/bxd72ph0ujt5b6b/DB_lockup_crash_and_exceptions_3_20.zip?dl=0

**Comment by metachronism** [comment]
Fixed in 7.3!
